MDEV-10814: innodb large allocations - Don't dump

grooverdan opened the pull request:
Rework of #333. This is hooks allocate_large which is used for innodb buffer pool allocation and parts of online table operations.

An alternate PR covering just the innodb buffer pool pages is also provided if this is too general. I prefer this implementation.

I submit this under the MCA.

svoj commented:
Hi Daniel,

Thanks for updating this patch. Did you miss cmake checks for DONTDUMP?

Thanks,
Sergey

grooverdan commented:
aaae41519441c7f6a3f465972dbfe7468892b7be isn't part of the 10.2 branch it appears that isn't a detached commit floating around somewhere not associated with a branch. I agree it isn't desirable with this PR.

dr-m commented:
Thank you for the clarification. I prefer this #364 to #365.
I have one more request: Could we have a configuration parameter to control the core dump behaviour?

While most users of core dumps might have no use for the InnoDB buffer pool contents, it has been useful to me in a number of cases (usually involving deadlocks, corruption or race conditions). Often a crash would not reproduce after restarting the server. If the operation involves a write, usually the telltale redo log records would not yet have been flushed to the redo log files. (OK, the redo log buffer would still remain present in the core dump, but it would be much easier to access the buffer pool contents directly.)

svoj commented:
It was agreed in PR#333 that we don't add system variable:
* no system variables in first patch (we may have to add them in future though)

svoj commented:
What if we don't disable core dumps in debug version?

grooverdan commented:
There's a point at which in a debug build your are debugging a different product. A cmake variable around the madvise could minimise this a bit more but this still doesn't solve the needing the restart to to get a dump.

Alternative:

Is clearing private huge pages (bit 5) from the coredump using coredump_filter (ref: https://linux.die.net/man/5/core) too brutal? It can be changed at runtime or [gdb core generation can ignore it](https://sourceware.org/gdb/onlinedocs/gdb/Core-File-Generation.html)

I think the downside is some thp could end up being huge pages if an allocation falls into a large size and this might degrade other core dumps. Since MDEV-11059 I guess we don't need to worry about jemalloc variant.

Examination of huge page usage:
<pre>
sudo cat /proc/15348/smaps  | grep Huge | grep -v ' 0 kB'
</pre>

Systemd implementation:
<pre>
[dan@hopper rocksdb-build-script]$ systemctl status mariadb.service
● mariadb.service - MariaDB database server
   Loaded: loaded (/usr/lib/systemd/system/mariadb.service; disabled; vendor preset: disabled)
  Drop-In: /etc/systemd/system/mariadb.service.d
           └─coredump_filter.conf, migrated-from-my.cnf-settings.conf
   Active: active (running) since Tue 2017-08-29 09:39:03 AEST; 2s ago
  Process: 15378 ExecStartPost=/bin/bash -c 'echo $(( $(&lt;/proc/$MAINPID/coredump_filter) & ~(1 &lt;&lt; 4)  )) &gt;  /proc/$MAINPID/coredump_filter' (code=exited, status=0/SUCCESS)
  Process: 15376 ExecStartPost=/bin/sh -c systemctl unset-environment _WSREP_START_POSITION (code=exited, status=0/SUCCESS)
  Process: 15335 ExecStartPre=/bin/sh -c [ ! -e /usr/bin/galera_recovery ] && VAR= ||   VAR=`/usr/bin/galera_recovery`; [ $? -eq 0 ]   && systemctl set-environment _WSREP_START_POSITION=$VAR || exit 1 (co
  Process: 15333 ExecStartPre=/bin/sh -c systemctl unset-environment _WSREP_START_POSITION (code=exited, status=0/SUCCESS)
 Main PID: 15348 (mysqld)
   Status: "Taking your SQL requests now..."
    Tasks: 26 (limit: 4915)
   CGroup: /system.slice/mariadb.service
           └─15348 /usr/sbin/mysqld

Aug 29 09:39:03 hopper.ozlabs.ibm.com mysqld[15348]: 2017-08-29  9:39:03 140426190813504 [Note] InnoDB: Highest supported file format is Barracuda.
Aug 29 09:39:03 hopper.ozlabs.ibm.com mysqld[15348]: 2017-08-29  9:39:03 140426190813504 [Note] InnoDB: 128 rollback segment(s) are active.
Aug 29 09:39:03 hopper.ozlabs.ibm.com mysqld[15348]: 2017-08-29  9:39:03 140426190813504 [Note] InnoDB: Waiting for purge to start
Aug 29 09:39:03 hopper.ozlabs.ibm.com mysqld[15348]: 2017-08-29  9:39:03 140426190813504 [Note] InnoDB:  Percona XtraDB (http://www.percona.com) 5.6.36-82.1 started; log sequence number 1616839
Aug 29 09:39:03 hopper.ozlabs.ibm.com mysqld[15348]: 2017-08-29  9:39:03 140425432135424 [Note] InnoDB: Dumping buffer pool(s) not yet started
Aug 29 09:39:03 hopper.ozlabs.ibm.com mysqld[15348]: 2017-08-29  9:39:03 140426190813504 [Note] Plugin 'FEEDBACK' is disabled.
Aug 29 09:39:03 hopper.ozlabs.ibm.com mysqld[15348]: 2017-08-29  9:39:03 140426190813504 [Note] Server socket created on IP: '::'.
Aug 29 09:39:03 hopper.ozlabs.ibm.com mysqld[15348]: 2017-08-29  9:39:03 140426190813504 [Note] /usr/sbin/mysqld: ready for connections.
Aug 29 09:39:03 hopper.ozlabs.ibm.com mysqld[15348]: Version: '10.1.26-MariaDB'  socket: '/var/lib/mysql/mysql.sock'  port: 3306  MariaDB Server
Aug 29 09:39:03 hopper.ozlabs.ibm.com systemd[1]: Started MariaDB database server.
[dan@hopper rocksdb-build-script]$ cat /proc/15348/coredump_filter
0000000b
[dan@hopper rocksdb-build-script]$ cat /etc/systemd/system/mariadb.service.d/coredump_filter.conf

[Service]
ExecStartPost=/bin/bash -c 'echo $(( $(&lt;/proc/$MAINPID/coredump_filter) & ~(1 &lt;&lt; 4)  )) &gt;  /proc/$MAINPID/coredump_filter'
</pre>

dr-m commented:
Finally, I had a short discussion over this with @svoj tonight, and I think that his solution from August 28 makes sense.

The InnoDB buffer pool is often "polluting" the core dumps and making them unusably big on production systems, or by including too much confidential data.
For the few rare cases where the core dump would be useful for debugging an InnoDB bug or data corruption, a debug server could indeed be used.

So, how about the following? Disable the core dump for the buffer pool (and possible other structures, such as the log_sys and recv_sys buffers) on the non-debug server, but keep it enabled for the non-debug server.

Also, earlier today @montywi mentioned some tweak that he uses for preventing the buffer pool from being included in the core dumps. If it is something that can be disabled or enabled at runtime, that could be the desired solution: disable initially, and have some way of re-enabling the full core dump after the server process has been started.

grooverdan commented:
I suspect Monty's tweak might of the the coredump filter mentioned earlier (and RFE for systemd in https://github.com/systemd/systemd/issues/6685). Its a large scope but script-able.

log_t->buf{,_ptr) using log_buffer_size(*2) seems worth including.
recv_sys->buf  seems limited to 2M (RECV_PARSING_BUF_SIZE) which doesn't seems worth worrying about from a size perspective. If easy to handle will include this from a confidentiality point of view.

I'll include a function that marked with visibility default that make these sections are dumpable so this can be called in a gdb script before the core dump if needed.

grooverdan commented:
Updated as requested. log and recv needed to be moved to large memory pool (otherwise madvise would EINVAL on the address). It seemed prudent to make these large memories anyway.

Below is a manual test case showing that the madvise call was called on the innodb buffer pool, followed by the log, followed by recv. Recv buffer is freed after the start up recovery. Calling buf_madvise_do_dump from gdb returns 0 indicating all madvise(DO_DUMP) calls where successful.

( cd mysql-test/; ./mtr --mem --manual-gdb  --port-base=16000  --max-test-fail=40 innodb_buffer_pool_load_at_startup_basic  )
<pre>
(gdb) break madvise
Breakpoint 1 at 0x5516d0
(gdb) break do_handle_one_connection(CONNECT*) 
Breakpoint 2 at 0x6c571f: file /home/dan/repos/mariadb-server-10.3/sql/sql_connect.cc, line 1359.
(gdb) run
Starting program: /home/dan/repos/build-mariadb-10.3/sql/mysqld --defaults-group-suffix=.1 --defaults-file=/home/dan/repos/build-mariadb-10.3/mysql-test/var/my.cnf --log-output=file --gdb --innodb --innodb-cmpmem --innodb-cmp-per-index --innodb-trx --innodb-locks --innodb-metrics --innodb-buffer-pool-stats --innodb-buffer-page --innodb-buffer-page-lru --innodb-sys-columns --innodb-sys-fields --innodb-sys-foreign --innodb-sys-foreign-cols --innodb-sys-indexes --innodb-sys-tables --innodb-sys-virtual --core-file --loose-debug-sync-timeout=300 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
2017-11-20 21:25:43 0 [Note] /home/dan/repos/build-mariadb-10.3/sql/mysqld (mysqld 10.3.3-MariaDB-log) starting as process 9884 ...
[New Thread 0x7ffff016a700 (LWP 9888)]
[New Thread 0x7fffeedfe700 (LWP 9889)]

Thread 1 "mysqld" hit Breakpoint 1, madvise () at ../sysdeps/unix/syscall-template.S:84
84      T_PSEUDO (SYSCALL_SYMBOL, SYSCALL_NAME, SYSCALL_NARGS)
(gdb) bt
#0  madvise () at ../sysdeps/unix/syscall-template.S:84
#1  0x0000000000a87b2b in ut_allocator<unsigned char, true>::allocate_large (this=<optimized out>, n_elements=<optimized out>, pfx=0x21d98a0, 
    set_to_zero=<error reading variable: access outside bounds of object referenced via synthetic pointer>, dontdump=<optimized out>)
    at /home/dan/repos/mariadb-server-10.3/storage/innobase/include/ut0new.h:617
#2  0x0000000000a87239 in buf_chunk_init (buf_pool=0x21d91d0, chunk=0x21d9890, mem_size=8601600)
    at /home/dan/repos/mariadb-server-10.3/storage/innobase/buf/buf0buf.cc:1552
#3  0x0000000000a7803e in buf_pool_init_instance (buf_pool=<optimized out>, buf_pool_size=<optimized out>, instance_no=<optimized out>)
    at /home/dan/repos/mariadb-server-10.3/storage/innobase/buf/buf0buf.cc:1834
#4  buf_pool_init (total_size=<optimized out>, n_instances=<optimized out>)
    at /home/dan/repos/mariadb-server-10.3/storage/innobase/buf/buf0buf.cc:2060
#5  0x0000000000a12917 in innobase_start_or_create_for_mysql () at /home/dan/repos/mariadb-server-10.3/storage/innobase/srv/srv0start.cc:1828
#6  0x000000000093ca48 in innobase_init (p=<optimized out>) at /home/dan/repos/mariadb-server-10.3/storage/innobase/handler/ha_innodb.cc:4147
#7  0x00000000007b7b43 in ha_initialize_handlerton (plugin=0x1f7f5e0) at /home/dan/repos/mariadb-server-10.3/sql/handler.cc:521
#8  0x000000000060c3eb in plugin_initialize (tmp_root=<optimized out>, plugin=0x1f7f5e0, argc=<optimized out>, argv=<optimized out>, 
    options_only=<error reading variable: access outside bounds of object referenced via synthetic pointer>)
    at /home/dan/repos/mariadb-server-10.3/sql/sql_plugin.cc:1438
#9  0x000000000060c193 in plugin_init (argc=0x15a8258 <remaining_argc>, argv=0x1e44c50, flags=2)
    at /home/dan/repos/mariadb-server-10.3/sql/sql_plugin.cc:1720
#10 0x0000000000571731 in init_server_components () at /home/dan/repos/mariadb-server-10.3/sql/mysqld.cc:5340
#11 0x000000000056d558 in mysqld_main (argc=<optimized out>, argv=<optimized out>) at /home/dan/repos/mariadb-server-10.3/sql/mysqld.cc:5935
#12 0x00007ffff55a488a in __libc_start_main (main=0x569520 <main(int, char**)>, argc=23, argv=0x7fffffffd218, init=<optimized out>, 
    fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7fffffffd208) at ../csu/libc-start.c:295
#13 0x000000000056946a in _start ()
(gdb) c
Continuing.

Thread 1 "mysqld" hit Breakpoint 1, madvise () at ../sysdeps/unix/syscall-template.S:84
84      T_PSEUDO (SYSCALL_SYMBOL, SYSCALL_NAME, SYSCALL_NARGS)
(gdb) bt
#0  madvise () at ../sysdeps/unix/syscall-template.S:84
#1  0x000000000097d9ad in ut_zalloc_dontdump (n_bytes=2101248, key=0x2255560)
    at /home/dan/repos/mariadb-server-10.3/storage/innobase/include/ut0new.h:973
#2  0x000000000097ec6b in log_sys_init () at /home/dan/repos/mariadb-server-10.3/storage/innobase/log/log0log.cc:729
#3  0x0000000000a12972 in innobase_start_or_create_for_mysql () at /home/dan/repos/mariadb-server-10.3/storage/innobase/srv/srv0start.cc:1852
#4  0x000000000093ca48 in innobase_init (p=<optimized out>) at /home/dan/repos/mariadb-server-10.3/storage/innobase/handler/ha_innodb.cc:4147
#5  0x00000000007b7b43 in ha_initialize_handlerton (plugin=0x1f7f5e0) at /home/dan/repos/mariadb-server-10.3/sql/handler.cc:521
#6  0x000000000060c3eb in plugin_initialize (tmp_root=<optimized out>, plugin=0x1f7f5e0, argc=<optimized out>, argv=<optimized out>, 
    options_only=<error reading variable: access outside bounds of object referenced via synthetic pointer>)
    at /home/dan/repos/mariadb-server-10.3/sql/sql_plugin.cc:1438
#7  0x000000000060c193 in plugin_init (argc=0x15a8258 <remaining_argc>, argv=0x1e44c50, flags=2)
    at /home/dan/repos/mariadb-server-10.3/sql/sql_plugin.cc:1720
#8  0x0000000000571731 in init_server_components () at /home/dan/repos/mariadb-server-10.3/sql/mysqld.cc:5340
#9  0x000000000056d558 in mysqld_main (argc=<optimized out>, argv=<optimized out>) at /home/dan/repos/mariadb-server-10.3/sql/mysqld.cc:5935
#10 0x00007ffff55a488a in __libc_start_main (main=0x569520 <main(int, char**)>, argc=23, argv=0x7fffffffd218, init=<optimized out>, 
    fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7fffffffd208) at ../csu/libc-start.c:295
#11 0x000000000056946a in _start ()
(gdb) c
Continuing.

Thread 1 "mysqld" hit Breakpoint 1, madvise () at ../sysdeps/unix/syscall-template.S:84
84      T_PSEUDO (SYSCALL_SYMBOL, SYSCALL_NAME, SYSCALL_NARGS)
(gdb) bt
#0  madvise () at ../sysdeps/unix/syscall-template.S:84
#1  0x0000000000982907 in ut_malloc_dontdump (n_bytes=2097152, key=<optimized out>)
    at /home/dan/repos/mariadb-server-10.3/storage/innobase/include/ut0new.h:953
#2  recv_sys_init () at /home/dan/repos/mariadb-server-10.3/storage/innobase/log/log0recv.cc:556
#3  0x0000000000a12977 in innobase_start_or_create_for_mysql () at /home/dan/repos/mariadb-server-10.3/storage/innobase/srv/srv0start.cc:1854
#4  0x000000000093ca48 in innobase_init (p=<optimized out>) at /home/dan/repos/mariadb-server-10.3/storage/innobase/handler/ha_innodb.cc:4147
#5  0x00000000007b7b43 in ha_initialize_handlerton (plugin=0x1f7f5e0) at /home/dan/repos/mariadb-server-10.3/sql/handler.cc:521
#6  0x000000000060c3eb in plugin_initialize (tmp_root=<optimized out>, plugin=0x1f7f5e0, argc=<optimized out>, argv=<optimized out>, 
    options_only=<error reading variable: access outside bounds of object referenced via synthetic pointer>)
    at /home/dan/repos/mariadb-server-10.3/sql/sql_plugin.cc:1438
#7  0x000000000060c193 in plugin_init (argc=0x15a8258 <remaining_argc>, argv=0x1e44c50, flags=2)
    at /home/dan/repos/mariadb-server-10.3/sql/sql_plugin.cc:1720
#8  0x0000000000571731 in init_server_components () at /home/dan/repos/mariadb-server-10.3/sql/mysqld.cc:5340
#9  0x000000000056d558 in mysqld_main (argc=<optimized out>, argv=<optimized out>) at /home/dan/repos/mariadb-server-10.3/sql/mysqld.cc:5935
#10 0x00007ffff55a488a in __libc_start_main (main=0x569520 <main(int, char**)>, argc=23, argv=0x7fffffffd218, init=<optimized out>, 
    fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7fffffffd208) at ../csu/libc-start.c:295
#11 0x000000000056946a in _start ()
(gdb) c
Continuing.
[New Thread 0x7fffed5c7700 (LWP 9890)]
[New Thread 0x7fffecdc6700 (LWP 9891)]
[New Thread 0x7fffe7fff700 (LWP 9892)]
[New Thread 0x7fffe77fe700 (LWP 9893)]
[New Thread 0x7fffe6ffd700 (LWP 9894)]
[New Thread 0x7fffe67fc700 (LWP 9895)]
[New Thread 0x7fffe5ffb700 (LWP 9896)]

Thread 1 "mysqld" hit Breakpoint 1, madvise () at ../sysdeps/unix/syscall-template.S:84
84      T_PSEUDO (SYSCALL_SYMBOL, SYSCALL_NAME, SYSCALL_NARGS)
(gdb) bt
#0  madvise () at ../sysdeps/unix/syscall-template.S:84
#1  0x0000000000982b42 in ut_free_dodump (ptr=0x7fffed5c8000, key=<optimized out>)
    at /home/dan/repos/mariadb-server-10.3/storage/innobase/include/ut0new.h:997
#2  recv_sys_debug_free () at /home/dan/repos/mariadb-server-10.3/storage/innobase/log/log0recv.cc:591
#3  0x0000000000989458 in recv_recovery_from_checkpoint_finish () at /home/dan/repos/mariadb-server-10.3/storage/innobase/log/log0recv.cc:3436
#4  0x0000000000a142f3 in innobase_start_or_create_for_mysql () at /home/dan/repos/mariadb-server-10.3/storage/innobase/srv/srv0start.cc:2312
#5  0x000000000093ca48 in innobase_init (p=<optimized out>) at /home/dan/repos/mariadb-server-10.3/storage/innobase/handler/ha_innodb.cc:4147
#6  0x00000000007b7b43 in ha_initialize_handlerton (plugin=0x1f7f5e0) at /home/dan/repos/mariadb-server-10.3/sql/handler.cc:521
#7  0x000000000060c3eb in plugin_initialize (tmp_root=<optimized out>, plugin=0x1f7f5e0, argc=<optimized out>, argv=<optimized out>, 
    options_only=<error reading variable: access outside bounds of object referenced via synthetic pointer>)
    at /home/dan/repos/mariadb-server-10.3/sql/sql_plugin.cc:1438
#8  0x000000000060c193 in plugin_init (argc=0x15a8258 <remaining_argc>, argv=0x1e44c50, flags=2)
    at /home/dan/repos/mariadb-server-10.3/sql/sql_plugin.cc:1720
#9  0x0000000000571731 in init_server_components () at /home/dan/repos/mariadb-server-10.3/sql/mysqld.cc:5340
#10 0x000000000056d558 in mysqld_main (argc=<optimized out>, argv=<optimized out>) at /home/dan/repos/mariadb-server-10.3/sql/mysqld.cc:5935
#11 0x00007ffff55a488a in __libc_start_main (main=0x569520 <main(int, char**)>, argc=23, argv=0x7fffffffd218, init=<optimized out>, 
    fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7fffffffd208) at ../csu/libc-start.c:295
#12 0x000000000056946a in _start ()
(gdb) c
Continuing.
[New Thread 0x7fffe57fa700 (LWP 9898)]
[New Thread 0x7fffe4ff9700 (LWP 9899)]
[New Thread 0x7fffdbfff700 (LWP 9900)]
[Switching to Thread 0x7fffe57fa700 (LWP 9898)]

Thread 11 "mysqld" hit Breakpoint 1, madvise () at ../sysdeps/unix/syscall-template.S:84
84      T_PSEUDO (SYSCALL_SYMBOL, SYSCALL_NAME, SYSCALL_NARGS)
(gdb) bt
#0  madvise () at ../sysdeps/unix/syscall-template.S:84
#1  0x00007ffff7bbe628 in start_thread (arg=0x7fffe57fa700) at pthread_create.c:556
#2  0x00007ffff5694e1f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:97
(gdb) c
Continuing.
[New Thread 0x7fffdb7fe700 (LWP 9901)]
[New Thread 0x7fffdaffd700 (LWP 9902)]
[Thread 0x7fffe57fa700 (LWP 9898) exited]
[New Thread 0x7fffe57fa700 (LWP 9903)]
[New Thread 0x7fffda7fc700 (LWP 9904)]
[New Thread 0x7fffd9ffb700 (LWP 9905)]
[New Thread 0x7fffd97fa700 (LWP 9906)]
[New Thread 0x7fffd8ff9700 (LWP 9907)]
[New Thread 0x7fffc7fff700 (LWP 9908)]
[New Thread 0x7fffc77fe700 (LWP 9909)]
[New Thread 0x7fffc6ffd700 (LWP 9910)]
[New Thread 0x7fffc67fc700 (LWP 9911)]
[New Thread 0x7fffc5ffb700 (LWP 9912)]
[New Thread 0x7fffed7c1700 (LWP 9913)]
[New Thread 0x7fffed777700 (LWP 9914)]
[New Thread 0x7fffed72d700 (LWP 9917)]
[Switching to Thread 0x7fffed72d700 (LWP 9917)]

Thread 28 "mysqld" hit Breakpoint 2, do_handle_one_connection (connect=0x238d1c8) at /home/dan/repos/mariadb-server-10.3/sql/sql_connect.cc:1359
1359      ulonglong thr_create_utime= microsecond_interval_timer();
(gdb) clear madvise
Deleted breakpoint 1
(gdb) call buf_madvise_do_dump()
$1 = 0
(gdb) c
Continuing.
</pre>

ari commented:
I'd like to add another use-case for this patch... when a galera cluster node dies, galera doesn't seem to detect that node has gone away until the core dump completes. On our systems that takes many minutes, during which time the cluster is non-functional because replication data is still trying to be sent to the node which is non-responsive but hasn't fully died.

Granted mariaDB shouldn't core dump that often, but with MDEV-12837 it has been a big problem.

dr-m commented on storage/innobase/include/ut0new.h:
```diff
@@ -843,10 +886,18 @@ ut_delete_array(
+#define ut_zalloc_dontdump(n_bytes, key) static_cast<void*>( \
+	ut_allocator<byte>(key).allocate_large( \
+		n_bytes, true, true))
+
```
I don’t think we really need `ut_zalloc_dontdump()`.

dr-m commented on storage/innobase/include/ut0new.h:
```diff
@@ -568,14 +584,21 @@ class ut_allocator {
+	initialized with 0x0 bytes.
+	@param[in]      dontdump        if true, advise the OS is not to core
+	dump this memory.
+        The caller must provide space for this one and keep
```
Indent the last comment line by TAB, not 8 spaces.

dr-m commented on storage/innobase/include/ut0new.h:
```diff
@@ -586,13 +609,23 @@ class ut_allocator {
+                {
+			ib::warn() << "Failed to set memory to DONTDUMP: "
+                                                                << strerror(errno);
+                }
```
Could you please define a common utility function for the ‘dontdump’ and ‘dodump’ calls? These calls are duplicated at least in 2 places.

grooverdan commented on storage/innobase/include/ut0new.h:
```diff
@@ -888,14 +943,69 @@ ut_delete_array(
+#ifdef UNIV_PFS_MEMORY
+		allocate_trace(n_bytes, NULL, key);
+#endif
+		::memset(ptr, 1, n_bytes);
```
I checked and shmget(IPC_PRIVATE), mmap (MAP_PRIVATE) and VirtualAlloc provide initialised memory.

grooverdan commented on storage/innobase/include/log0recv.h:
```diff
@@ -215,6 +215,9 @@ struct recv_sys_t{
+	ut_new_pfx_t	mem_pfx;/*!< Auxiliary structure, for perfomance
+				schema memory - buf_ptr */
+				/*!< buf size in bytes */
 	byte*		buf;	/*!< buffer for parsing log records */
```
I'm assuming this is just a comment as it currently isn't. Fixed the buf_ptr reference in the comment.

dr-m replied:
It was a suggestion to remove `recv_sys_t::buf_ptr` and to directly allocate aligned memory in `recv_sys_t::buf`.
On a second reading, I noticed the stray comment `/*!< buf size in bytes` which should be  removed.

grooverdan commented on storage/innobase/include/log0log.h:
```diff
@@ -612,6 +612,8 @@ struct log_t{
+					schema memory - buf_ptr */
 	byte*		buf_ptr;	/*!< unaligned log buffer, which should
 					be of double of buf_size */
 	byte*		buf;		/*!< log buffer currently in use;
```
ok done in 3bb5bc3. There may be some more ut_align / OS_FILE_LOG_BLOCK_SIZE things that can be simplified however I've kept this minimal for now.

dr-m commented on storage/innobase/include/ut0new.h:
```diff
@@ -568,14 +619,18 @@ class ut_allocator {
+	allocated memory.
+	@param[in]      dontdump        if true, advise the OS is not to core
+	dump this memory.
+        The caller must provide space for this one and keep
```
I believe that the `@param[in] dontdump` line should be added immediately before the `@return`. The comment of `pfx` is several lines long.

dr-m commented on storage/innobase/include/ut0new.h:
```diff
@@ -235,6 +239,45 @@ struct ut_new_pfx_t {
+	}
+}
+
+static void ut_dodump(void* ptr, ut_new_pfx_t* pfx)
```
`pfx` should be passed as pointer to `const`

dr-m commented on storage/innobase/include/ut0new.h:
```diff
@@ -357,13 +404,17 @@ class ut_allocator {
 
 		return(reinterpret_cast<pointer>(pfx + 1));
 #else
+		ut_dontdump(ptr, total_bytes, dontdump, NULL, file);
```
If we call `ut_dontdump()` with `pfx=NULL`, how can the `ut_dodump()` be called when the memory is freed? `ut_dodump()` always relies on `pfx` to determine the size.

dr-m commented on storage/innobase/include/ut0new.h:
```diff
@@ -896,6 +968,18 @@ ut_delete_array(
+#endif
+	ut_dodump(ptr, key);
+
+	os_mem_free_large(ptr, size);
```
Why do we need to assign `size = key->m_size`? If `deallocate_trace(key)` invalidates `key->m_size`, then already the call `ut_dodump(ptr, key)` would be using an invalid `key`.

grooverdan commented on storage/innobase/buf/buf0buf.cc:
```diff
@@ -1173,6 +1173,57 @@ buf_page_is_corrupted(
+		chunk = buf_pool->chunks;
+
+		for (int n = buf_pool->n_chunks; n--; chunk++)
+		{
```
er, there isn't. Just spaced big on github I guess.

vuvova commented:
auto-closed by mistake (one of pushed commits had #364 in the comment)

grooverdan commented:
No problem. haven't forgotten about one. Just haven't had time to correct according to review comments yet. Sorry for the delay.

grooverdan commented:
Finally completed the rework of this.

Used size fields rather than relying on pfx structures. As such it looks a lot simpler, even if there is a size variable being maintained for each large allocation where advice dontdump.

```row0log.cc``` / ```row0merge.cc```, despite the modification isn't changed (default to allocate_large is ```dontdump=false``), its just tracking the pool size for a consistent API. Also conveniently ensures that the crypt block is the same size as the other block.

grooverdan commented:
thanks @dr-m, sorry it took so long to get back to this.
